homeassistant/helpers/service_platform.py

```diff
@@ -0,0 +1,526 @@
+            return self.parallel_updates
+
+        self.parallel_updates_created = True
+
+        parallel_updates = getattr(self.platform, "PARALLEL_UPDATES", None)
+
+        if parallel_updates is None and not service_is_async:
+            parallel_updates = 1
+
+        if parallel_updates == 0:
+            parallel_updates = None
+
+        if parallel_updates is not None:
+            self.parallel_updates = asyncio.Semaphore(parallel_updates)
+
+        return self.parallel_updates
+
+    @callback
+    def async_shutdown(self) -> None:
+        """Call when Home Assistant is stopping."""
+        self.async_cancel_retry_setup()
+
+    @callback
+    def async_cancel_retry_setup(self) -> None:
+        """Cancel retry setup."""
+        if self._async_cancel_retry_setup is not None:
+            self._async_cancel_retry_setup()
+            self._async_cancel_retry_setup = None
+
+    async def async_setup_entry(self, config_entry: ConfigEntry) -> bool:
+        """Set up the platform from a config entry."""
+        self.config_entry = config_entry
+        platform = self.platform
+
+        @callback
+        def async_create_setup_task() -> Coroutine:
+            """Get task to set up platform."""
+            current_entry.set(config_entry)
+            return platform.async_setup_entry(
+                self.hass, config_entry, self._async_schedule_add_services
+            )
+
+        return await self._async_setup_platform(async_create_setup_task)
+
+    async def _async_setup_platform(
+        self, async_create_setup_task: Callable[[], Coroutine], tries: int = 0
+    ) -> bool:
+        """Set up a platform via config file or config entry.
+
+        async_create_setup_task creates a coroutine that sets up platform.
+        """
+        logger = self.logger
+        hass = self.hass
+        full_name = f"{self.domain}.{self.platform_name}"
+
+        logger.info("Setting up %s", full_name)
+        warn_task = hass.loop.call_later(
+            SLOW_SETUP_WARNING,
+            logger.warning,
+            "Setup of %s platform %s is taking over %s seconds.",
+            self.domain,
+            self.platform_name,
+            SLOW_SETUP_WARNING,
+        )
+        with async_start_setup(hass, [full_name]):
+            try:
+                task = async_create_setup_task()
+
+                async with hass.timeout.async_timeout(SLOW_SETUP_MAX_WAIT, self.domain):
+                    await asyncio.shield(task)
+
+                # Block till all services are added
+                while self._tasks:
+                    pending = [task for task in self._tasks if not task.done()]
+                    self._tasks.clear()
+
+                    if pending:
+                        await asyncio.gather(*pending)
+
+                hass.config.components.add(full_name)
+                self._setup_complete = True
+                return True
+            except PlatformNotReady as ex:
```

Comment:
Let’s not support PlatformNotReady. I think that we should model the Service Platform around `async_setup_entry`, and I _think_ that we don’t support that one there?

Reply:
We do support it there. That's why I included it.